src/mjlab/utils/buffers/delay_buffer.py

```diff
@@ -0,0 +1,267 @@
+    """Current lag per environment. Shape: (batch_size,)."""
+    return self._current_lags.clone()
+
+  def reset(self, batch_ids: Sequence[int] | torch.Tensor | None = None) -> None:
+    """Reset specified environments to initial state.
+
+    Args:
+      batch_ids: Batch indices to reset, or None to reset all.
+    """
+    ids = slice(None) if batch_ids is None else batch_ids
+    self._buffer.reset(batch_ids=batch_ids)
+    self._current_lags[ids] = 0
+    self._step_count[ids] = 0
+
+    if self.update_period > 0 and self.per_env_phase:
+      if batch_ids is None:
+        self._phase_offsets = torch.randint(
+          0,
+          self.update_period,
+          (self.batch_size,),
+          dtype=torch.long,
+          device=self.device,
+          generator=self.generator,
+        )
+      else:
+        new_phases = torch.randint(
+          0,
+          self.update_period,
+          (self.batch_size,),
+          dtype=torch.long,
+          device=self.device,
+          generator=self.generator,
+        )
+        self._phase_offsets[ids] = new_phases[ids]
+
+  def append(self, data: torch.Tensor) -> None:
+    """Append new observation to buffer.
+
+    Args:
+      data: Observation tensor of shape (batch_size, ...).
+    """
+    self._buffer.append(data)
+
+  def compute(self) -> torch.Tensor:
+    """Compute delayed observation for current step.
+
+    Returns:
+      Delayed observation with shape (batch_size, ...).
+    """
+    if not self.is_initialized:
+      raise RuntimeError("Buffer not initialized. Call append() first.")
+
+    self._update_lags()
+
+    # Clamp lags to valid range [0, buffer_length - 1].
+    # Buffer may not be full yet (e.g., only 2 frames but sampled lag=3).
+    valid_lags = torch.minimum(self._current_lags, self._buffer.current_length - 1)
+    valid_lags = torch.maximum(valid_lags, torch.zeros_like(valid_lags))
+
+    return self._buffer[valid_lags]
+
+  def _update_lags(self) -> None:
+    """Update current lags according to configured policy."""
+    if self.update_period > 0:
+      phase_adjusted_count = (self._step_count + self._phase_offsets) % (
+        self.update_period
+      )
+      should_update = phase_adjusted_count == 0
+    else:
+      should_update = torch.ones(self.batch_size, dtype=torch.bool, device=self.device)
+
+    if torch.any(should_update):
+      new_lags = self._sample_lags(should_update)
+      self._current_lags = torch.where(should_update, new_lags, self._current_lags)
+
+    self._step_count += 1
+
+  def _sample_lags(self, mask: torch.Tensor) -> torch.Tensor:
+    """Sample new lags for specified environments.
+
+    Args:
+      mask: Boolean mask of shape (batch_size,) indicating which envs to sample.
+
+    Returns:
+      New lags with shape (batch_size,).
+    """
+    new_lags = self._current_lags.clone()
+
+    if not torch.any(mask):
+      return new_lags
+
+    num_to_sample = int(mask.sum().item())
```

Comment:
possible performance concern: it feels like this method will introduce a lot of CUDA syncs from the `.item()` calls and conditioning on `torch.any()`, is this a concern?

I'm curious if a branch-free / async-friendly implementation would be faster. Maybe there's similar optimizations that can be made in other parts of the codebase?